snuba/datasets/table_storage.py

```diff
@@ -33,12 +34,29 @@ def __init__(
         commit_log_topic: Optional[str] = None,
     ) -> None:
         self.__processor = processor
-        self.__default_topic_spec = KafkaTopicSpec(topic_name=default_topic)
+        self.__default_topic_spec = KafkaTopicSpec(
+            topic_name=default_topic,
+            partitions_number=settings.TOPIC_PARTITION_COUNTS.get(default_topic, 1),
+        )
         self.__replacement_topic_spec = (
-            KafkaTopicSpec(topic_name=replacement_topic) if replacement_topic else None
+            KafkaTopicSpec(
+                topic_name=replacement_topic,
+                partitions_number=settings.TOPIC_PARTITION_COUNTS.get(
+                    replacement_topic, 1
+                ),
+            )
+            if replacement_topic
+            else None
         )
         self.__commit_log_topic_spec = (
-            KafkaTopicSpec(topic_name=commit_log_topic) if commit_log_topic else None
+            KafkaTopicSpec(
+                topic_name=commit_log_topic,
+                partitions_number=settings.TOPIC_PARTITION_COUNTS.get(
+                    commit_log_topic, 1
+                ),
+            )
+            if commit_log_topic
+            else None
```

Comment:
I kind of hate this but I don't have any better ideas (at least that can be done relatively quickly.)

This needs to be centralized with where the topics are defined — doing it in the dataset definition doesn't make much sense, since two datasets can reference the same topic with potentially conflicting partition counts.

In my opinion, we also should start using some sort of tooling to build topics based on these configurations in production instead of doing everything with ad hoc configuration. This also means that if topic configuration changes, we'll also need some sort of migration tooling.

Ideally we'd be able to warn (or potentially even throw) if this definition doesn't reflect the actual configuration in production, since a mismatch here could cause subscriptions to be erroneously routed, etc.

Reply:
I mentioned something like this in the related PR. The way KafkaTopicSpec was defined here was because it should be used when creating the topic, otherwise Snuba never had the need to load parameters like the partition count. 
When we start needing the count to make our system works, I think we should not have two configs (sentry and snuba) that are independent but with an implicit dependency for them to be consistent.

Do you think it would be feasible to make Snuba get the partition count for an existing topic directly from Kafka and skipping the configuration ? So Snuba adapts itself to the way the topic is set.
This way this KafkaTopicSpec is, like before, used only to set up topics in development environments.

Reply:
I posted this comment on the other PR right before reading this, and I think opinion here still holds: https://github.com/getsentry/snuba/pull/746#discussion_r374439183

> I think it would make sense to warn loudly if the configured value doesn't match the actual value but I'd be very hesitant to dynamically change this divisor.

I think in the worst case scenario where this number changes without any corresponding data migration, I'd rather the data be consistently wrong (everything using the old divisor) rather than an unknown quantity of data being wrong (partitioned using the old divisor) and an unknown quantity of data being right (partitioned using the new divisor.)

Reply:
What about preventing the consumer from starting if the config is different than what Kafka says ? Like we still rely on the config to know the quantity but we do not start if we know the quantity is wrong (at least during startup, of course this could change during the processing, but we may be able to catch that inconsistency during partition assignment/revoke)

Reply:
Yeah, I am fine with fail-stopping in this situation instead of just warning about it. We'll need to find some API to expose that though via the generic streams interface. I will take a look into doing that, but it may take some time. I'd prefer not to block this task on it, if possible.

It will be easier for the consumer to detect a topic resize/repartitioning on rebalance than the partitioner (#746.) The only reasonable approach that I can think of would be to check every N seconds/minutes that the partition count is what is expected, either synchronously on subscription creation, or asynchronously in a thread in the background.

Reply:
Yes, sure, we do not need to block on this.